docs/scalars.rst

```diff
@@ -0,0 +1,175 @@
+Scalars in graphql
+==================
+
+Scalars
+-------
+
+The leaf nodes of the graphql type system are called scalars.  Once you reach a scalar type you
+cannot descend down any further into the type hierarchy.  A scalar type is meant to represent
+an indivisible value.
+
+The graphql specification says that all implementations must have the following scalar types.
+
+* String aka ``GraphQLString`` - A UTF‐8 character sequence.
+* Boolean aka ``GraphQLBoolean`` - true or false.
+* Int aka ``GraphQLInt`` - A signed 32‐bit integer.
+* Float aka ``GraphQLFloat`` - A signed double-precision floating-point value.
+* ID aka ``GraphQLID`` - A unique identifier which is serialized in the same way as a String; however, defining it as an ID signifies that it is not intended to be human‐readable.
+
+graphql-java adds the following scalar types which are useful in Java based systems
+
+* Long aka ``GraphQLLong`` - a java.lang.Long based scalar
+* Short aka ``GraphQLShort`` - a java.lang.Short based scalar
+* Byte aka ``GraphQLByte``  - a java.lang.Byte based scalar
+* BigDecimal aka ``GraphQLBigDecimal`` - a java.math.BigDecimal based scalar
+* BigInteger aka ``GraphQLBigInteger`` - a java.math.BigInteger based scalar
+
+
+The class ``graphql.Scalars`` contains singleton instances of the provided scalar types
+
+Writing your Own Custom Scalars
+-------------------------------
+
+You can write your own custom scalar implementations.  In doing so you take on the responsibility for coercing values
+at runtime, which we will explain in a moment.
+
+Imagine we decide we need to have an email scalar type.  It will take email addresses as input and output.
+
+We would create a singleton ``graphql.schema.GraphQLScalarType`` instance for this like so.
+
+.. code-block:: java
+
+        public static final GraphQLScalarType EMAIL = new GraphQLScalarType("email", "A custom scalar that handles emails", new Coercing() {
+            @Override
+            public Object serialize(Object dataFetcherResult) {
+                return serializeEmail(dataFetcherResult);
+            }
+
+            @Override
+            public Object parseValue(Object input) {
+                return parseEmailFromVariable(input);
+            }
+
+            @Override
+            public Object parseLiteral(Object input) {
+                return parseEmailFromAstLiteral(input);
+            }
+        });
+
+
+Coercing values
+---------------
+
+The real work in any custom scalar implementation is the ``graphql.schema.Coercing`` implementation.  This is responsible for 3 functions
+
+* ``parseValue`` - takes a variable input object and converts into the Java runtime representation
+* ``parseLiteral`` - takes an AST literal ``graphql.language.Value` as input and converts into the Java runtime representation
```

Comment:
this only happens when the scalar value is defined as a literal as part of the query (as opposed to using a variable)